core/src/main/java/com/linecorp/armeria/server/ContextPathDecoratingBindingBuilder.java

```diff
@@ -0,0 +1,244 @@
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> matchesParams(String... paramPredicates) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.matchesParams(paramPredicates);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> matchesParams(Iterable<String> paramPredicates) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.matchesParams(paramPredicates);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> matchesParams(
+            String paramName, Predicate<? super String> valuePredicate) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.matchesParams(paramName, valuePredicate);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> matchesHeaders(String... headerPredicates) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.matchesHeaders(headerPredicates);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> matchesHeaders(Iterable<String> headerPredicates) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.matchesHeaders(headerPredicates);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> matchesHeaders(CharSequence headerName,
+                                                                 Predicate<? super String> valuePredicate) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.matchesHeaders(headerName, valuePredicate);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> addRoute(Route route) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.addRoute(route);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> exclude(String pathPattern) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.exclude(pathPattern);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ContextPathDecoratingBindingBuilder<T> exclude(Route excludedRoute) {
+        return (ContextPathDecoratingBindingBuilder<T>) super.exclude(excludedRoute);
+    }
+
+    /**
+     * Sets the {@code decorator} and returns the (ContextPathDecoratingBindingBuilder)
+     * that this {@link ContextPathDecoratingBindingBuilder} was created from.
+     *
+     * @param decorator the {@link Function} that decorates {@link HttpService}
+     */
+    public ContextPathServicesBuilder<T> build(
+            Function<? super HttpService, ? extends HttpService> decorator) {
+        requireNonNull(decorator, "decorator");
+        buildRouteList().forEach(
+                route -> contextPaths.forEach(contextPath -> builder.addRouteDecoratingService(
+                        new RouteDecoratingService(route, contextPath, decorator))));
```

Comment:
> when initializing this class and build a complete Route

This wasn't done intentionally because I preferred to pass the context path separately as a String.
From the pull request description:

> Added a contextPath parameter to ServiceConfigBuilder, RouteDecoratingService constructors so that we can check call paths using contextPath at compile-time. (and routes aren't double-applied with a prefix by mistake) 

Do you prefer that the prefix adding is done implicitly? (i.e. `ServiceConfigBuilder` constructor doesn't accept a `contextPath`)

Reply:
Another reason I wanted to add the `contextPath` parameter to `ServiceConfigBuilder` was because there are some code points that create a `ServiceConfigBuilder` directly and I didn't want to forget about these points.

e.g.
https://github.com/line/armeria/blob/7a92dd8fcda7452bd11c1c11e36b8abf92b2c67f/core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java#L169-L174

If we add to the route instead of `ServiceConfigBuilder`, then it is possible that we forget to add a prefix.

> add ContextPathRoute for type-safety and static analysis

If `ContextPathRoute` is added, I thought the above won't be possible, but let me know if I am misunderstanding your intention

Reply:
As of this PR, All `Route`s will have the default context path (`/`) or a custom context paht. `Route` is immutable so I thought `ContextPathRoute` could be used internally to hold the context path. `new ContextPathRoute(route, contextPath)`.

However, I don't see we implement `ContextPathRoute` here and I'm not confident if the suggestion is better. Let's keep it as is and discuss more when we really need it.